### PR TITLE
Support swap partitions and logical volumes in DiskCustomization (COMPOSER-2400)

### PR DESCRIPTION
**Pull request by achilleas-k**
This PR adds support for swap partitions and logical volumes when using the new `disk` blueprint customizations.

Internally, swap space is represented by a new `disk.Swap` type (a `disk.Element` type) which implements the `disk.PayloadEntity` interface.  I also added a new interface: `disk.FSTabEntity`.  This new entity variant is for representing items that can appear in the fstab file.  It differs from `disk.Mountable` in that it doesn't have a mountpoint, which makes it useful for distinguishing between filesystems and swap areas.  It will also be useful for defining swap files as part of the partitioning layout (even though it is not, strictly speaking, part of the partition table).

In the customizations, swap is represented by setting the filesystem type to "swap".  This is valid for both plain partitions and lvm logical volumes.  When the fs type is "swap", the mountpoint must always be empty.

### CODE REVIEW

File: pkg/disk/swap.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	return "swap"
+}
+
+func (s *Swap) Clone() Entity {
+	if s == nil {
+		return nil
+	}
+
+	return &Swap{
+		UUID:         s.UUID,
```
Comment (mvo5) [inline pkg/disk/swap.go:35]: Should we clone the label here as well?

File: pkg/disk/disk_test.go [hunk 2]
```diff
@@ -164,6 +164,7 @@ func TestCreatePartitionTable(t *testing.T) {
 			// /boot and subdirectories is exempt from this rule
 			return nil
 		}
+
```
Comment (mvo5) [inline pkg/disk/disk_test.go:168]: (nitpick) This change in the ` disk: add a test partition table with swap ` looks a bit unrelated(?)

File: pkg/osbuild/fstab_stage_test.go [hunk 3]
```diff
@@ -50,3 +53,77 @@ func TestAddFilesystem(t *testing.T) {
...
+			require := require.New(t)
+			pt := testdisk.TestPartitionTables[name]
+
+			// math/rand is good enough in this case
+			/* #nosec G404 */
+			rng := rand.New(rand.NewSource(0))
+			// populate UUIDs
+			pt.GenerateUUIDs(rng)
+
+			// print an informative failure message if a new test partition
```
Comment (mvo5) [inline pkg/osbuild/fstab_stage_test.go:117]: :heart:

File: pkg/disk/swap.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+		FSTabOptions: s.FSTabOptions,
+	}
```
Comment (thozza) [inline pkg/disk/swap.go:37]: `Label: s.Label,` is missing. Is this intentional? :thinking:

File: pkg/osbuild/mkswap_stage.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package osbuild
+
+type MkswapStageOptions struct {
+	UUID  string `json:"uuid"`
+	Label string `json:"label,omitempty"`
+}
+
+func (MkswapStageOptions) isStageOptions() {}
+
+func NewMkswapStage(options *MkswapStageOptions, devices map[string]Device) *Stage {
```
Comment (mvo5) [inline pkg/osbuild/mkswap_stage.go:10]: (niptick/followup) - I know its a may sound a bit silly but a tiny test like:
```go
package osbuild

import (
	"encoding/json"
	"testing"

	"github.com/stretchr/testify/assert"
)

var expectedJSON = `{
  "type": "org.osbuild.mkswap",
  "options": {
    "uuid": "8a1fc521-02a0-4917-92a9-90a44d7e6503",
    "label": "some-label"
  },
  "devices": {
    "root": {
      "type": "org.osbuild.loopback"
    }
  }
}`

func TestNewMkswapStage(t *testing.T) {
	devices := make(map[string]Device)
	devices["root"] = Device{
		Type: "org.osbuild.loopback",
	}

	options := MkswapStageOptions{
		UUID:  "8a1fc521-02a0-4917-92a9-90a44d7e6503",
		Label: "some-label",
	}
	stage := NewMkswapStage(&options, devices)
	b, err := json.MarshalIndent(stage, "", "  ")
	assert.NoError(t, err)
	assert.Equal(t, expectedJSON, string(b))
}

```  
(also in the tests-as-documentation sense)

File: pkg/disk/partition_table_test.go [hunk 6]
```diff
@@ -1259,6 +1259,105 @@ func TestNewCustomPartitionTable(t *testing.T) {
 				},
 			},
 		},
+		"plain+swap": {
```
Comment (mvo5) [inline pkg/disk/partition_table_test.go:1262]: :heart_eyes:

File: pkg/osbuild/mkfs_stages_test.go [hunk 7]
```diff
@@ -220,6 +220,88 @@ func TestGenFsStagesBtrfs(t *testing.T) {
 	}, stages)
 }
 
+func TestGenMkfsStagesLVM(t *testing.T) {
```
Comment (thozza) [inline pkg/osbuild/mkfs_stages_test.go:258]: It should probably be called `TestGenFsStagesLVM`. The commit message should be amended as well.

**Review by thozza — COMMENTED** [review]
Great work. I've found a few minor issues, but overall, this looks very nice :+1:

File: pkg/osbuild/mkfs_stage.go [hunk 8]
```diff
@@ -7,81 +7,96 @@ import (
...
+				UUID:  e.UUID,
+				Label: e.Label,
 			}
-			stage = NewMkfsBtrfsStage(options, stageDevices)
-		case "ext4":
-			options := &MkfsExt4StageOptions{
-				UUID:  fsSpec.UUID,
-				Label: fsSpec.Label,
+			stages = append(stages, NewMkfsBtrfsStage(options, stageDevices))
+			// Handle subvolumes here directly instead of collecting them in
```
Comment (mvo5) [inline pkg/osbuild/mkfs_stage.go:67]: Maybe add a line about the /why/ here as well? I assume just because of simplify? Or is there a deeper reason? Maybe `// Handle subvolumes directly here instead of in their own case because we can reuse the parent easier this way` (I totally made this up :)

**Review by mvo5 — APPROVED** [review]
Thank you!  This is very nice! I have a few inline suggestions/ideas/questions but I don't think there are any blockers, fine to merge to move this forward (or we can iterate here as you prefer).

A bit of a meta comment - splitting https://github.com/osbuild/images/pull/1072/commits/1f9e2a247b5874811ac6970686031d91bc9f1659 (the testdisk move) out beforehand would be something I would have preferred, it would be a very simple review as it just shuffles things around. Then this PR would have been smaller.

File: pkg/osbuild/mkfs_stage.go [hunk 9]
```diff
@@ -7,81 +7,96 @@ import (
...
 	stages := make([]*Stage, 0, len(pt.Partitions))
 
-	processedBtrfsPartitions := make(map[string]bool)
-	genStage := func(mnt disk.Mountable, path []disk.Entity) error {
-		t := mnt.GetFSType()
-		var stage *Stage
+	genStage := func(ent disk.Entity, path []disk.Entity) error {
+		switch e := ent.(type) {
+		case *disk.Filesystem:
+			// TODO: extract last device renaming into helper
```
Comment (mvo5) [inline pkg/osbuild/mkfs_stage.go:21]: Fwiw, I did a quick experiemnt on this in https://github.com/osbuild/images/commit/fa11558c23af57532a0b41d5f676726393a21843 - I would love to see this slightly cleaner (the linked patch is still not nice but avoids some of the repetition and is hopefully a starting point)

File: pkg/osbuild/mkfs_stages_test.go (hunk 7 above)
Comment (achilleas-k) [reply pkg/osbuild/mkfs_stages_test.go:258]: ah yes, rebase shenanigans.

File: pkg/osbuild/mkfs_stage.go (hunk 9 above)
Comment (achilleas-k) [reply pkg/osbuild/mkfs_stage.go:21]: I left this as a TODO because I didn't want to add extra things and it could use some investigation.  Sometimes we need to rename the leaf device and other times we don't, but maybe renaming it always is a good general approach and we can just add it tot he existing function.  Alternatively we can have a "rename last device" function.  Another alternative is to add an argument to the function to rename the last device to whatever the caller specifies.

I don't know which is cleaner, so I'd rather look into it separately :)

File: pkg/disk/swap.go (hunk 4 above)
Comment (achilleas-k) [reply pkg/disk/swap.go:37]: Not intentional.  Thanks for catching it.

File: pkg/osbuild/mkfs_stages_test.go (hunk 7 above)
Comment (achilleas-k) [reply pkg/osbuild/mkfs_stages_test.go:258]: Or more like "cherry-pick shenanigans".

File: pkg/disk/partition_table_test.go [hunk 10]
```diff
@@ -2130,3 +2282,23 @@ func TestNewCustomPartitionTableErrors(t *testing.T) {
...
 }
+
+func TestPartitionTableFeatures(t *testing.T) {
+	type testCase struct {
+		partitionType    string
+		expectedFeatures disk.PartitionTableFeatures
+	}
+	testCases := []testCase{
+		{"plain", disk.PartitionTableFeatures{XFS: true, FAT: true}},
+		{"plain-swap", disk.PartitionTableFeatures{XFS: true, FAT: true}},
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply pkg/disk/partition_table_test.go:2293]: Good idea.  A swap feature might come in handy.

File: pkg/disk/partition_table_test.go [hunk 11]
```diff
@@ -2130,3 +2282,23 @@ func TestNewCustomPartitionTableErrors(t *testing.T) {
...
+		{"luks", disk.PartitionTableFeatures{XFS: true, FAT: true, LUKS: true}},
+		{"luks+lvm", disk.PartitionTableFeatures{XFS: true, FAT: true, LUKS: true, LVM: true}},
+		{"btrfs", disk.PartitionTableFeatures{XFS: true, FAT: true, Btrfs: true}},
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply pkg/disk/partition_table_test.go:2296]: I don't know if it's surprising or not.  Depends on the understanding and meaning of "features" here, so I think it would be more appropriate to describe it in the `features()` function itself rather than the test.

File: pkg/disk/swap.go [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// For swap, the Freq and PassNo are always 0.
+func (s *Swap) GetFSTabOptions() (FSTabOptions, error) {
+	if s == nil {
+		return FSTabOptions{}, nil
+	}
+	return FSTabOptions{
+		MntOps: s.FSTabOptions,
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply pkg/disk/swap.go:66]: Not hard at all.

[the remark this answers is not in the document]
Comment (achilleas-k) [reply pkg/disk/swap.go:66]: Let's leave it for a follow-up though.

File: pkg/disk/swap.go (hunk 1 above)
Comment (achilleas-k) [reply pkg/disk/swap.go:35]: Yup.  My mistake.

File: pkg/disk/partition_table_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (achilleas-k) [reply pkg/disk/partition_table_test.go:2293]: I'll do the other bits in a follow-up (prepping now).

File: pkg/disk/partition_table_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (achilleas-k) [reply pkg/disk/partition_table_test.go:2296]: Will add to the features follow-up.

File: pkg/disk/disk_test.go (hunk 2 above)
Comment (achilleas-k) [reply pkg/disk/disk_test.go:168]: I disagree.  It's definitely completely unrelated.

File: pkg/osbuild/mkfs_stage.go (hunk 8 above)
Comment (achilleas-k) [reply pkg/osbuild/mkfs_stage.go:67]: Sure.

File: pkg/disk/swap.go (hunk 4 above)
Comment (achilleas-k) [reply pkg/disk/swap.go]: Fixed.

File: pkg/disk/btrfs.go [hunk 13]
```diff
@@ -155,6 +155,10 @@ func (bs *BtrfsSubvolume) GetMountpoint() string {
 	return bs.Mountpoint
 }
 
+func (bs *BtrfsSubvolume) GetFSFile() string {
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply pkg/disk/btrfs.go:158]: Done.

File: pkg/blueprint/disk_customizations.go [hunk 14]
```diff
@@ -450,6 +454,14 @@ var validPlainFSTypes = []string{
 }
 
 func (p *PartitionCustomization) validatePlain(mountpoints map[string]bool) error {
+	if p.FSType == "swap" {
+		// make sure the mountpoint is empty and return
+		if p.Mountpoint != "" {
+			return fmt.Errorf("mountpoint for swap partition must be empty")
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply pkg/blueprint/disk_customizations.go]: Done.

File: pkg/disk/filesystem.go [hunk 15]
```diff
@@ -55,6 +55,10 @@ func (fs *Filesystem) GetMountpoint() string {
 	return fs.Mountpoint
 }
 
+func (fs *Filesystem) GetFSFile() string {
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply pkg/disk/filesystem.go:58]: Done.

File: pkg/disk/partition_table.go [hunk 16]
```diff
@@ -590,6 +590,32 @@ func (pt *PartitionTable) ForEachMountable(cb MountableCallback) error {
 	return forEachMountable(pt, []Entity{pt}, cb)
 }
 
+type FSTabEntityCallback func(mnt FSTabEntity, path []Entity) error
+
+func forEachFSTabEntity(c Container, path []Entity, cb FSTabEntityCallback) error {
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply pkg/disk/partition_table.go:595]: Added (and prepped a similar test for `ForEachMountable()` for a follow-up).

File: pkg/disk/partition_table.go [hunk 17]
```diff
@@ -1254,25 +1280,44 @@ func addPlainPartition(pt *PartitionTable, partition blueprint.PartitionCustomiz
...
 			Label:        partition.Label,
 			Mountpoint:   partition.Mountpoint,
 			FSTabOptions: "defaults", // TODO: add customization
-		},
+		}
+	}
+
+	newpart := Partition{
+		Type:     partType,
+		Bootable: false,
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply pkg/disk/partition_table.go]: Done.

File: pkg/osbuild/mkswap_stage.go (hunk 5 above)
Comment (achilleas-k) [reply pkg/osbuild/mkswap_stage.go:10]: Done.

**Review by ondrejbudai — APPROVED** [review]
This is pure awesomeness, thank you! ❤️
